README
Test-WWW-Selenium is a perl driver and test library for Selenium Remote
Control (SRC).  You will need to setup a Selenium Server that can launch
browsers.

Selenium Remote Control (SRC) is a test tool that allows you to write
automated web application UI tests in any programming language against
any HTTP website using any mainstream JavaScript-enabled browser.  SRC
provides a Selenium Server, which can automatically start/stop/control
any supported browser. It works by using Selenium Core, a pure-HTML+JS
library that performs automated tasks in JavaScript; the Selenium
Server communicates directly with the browser using AJAX (XmlHttpRe-
quest).

This module sends commands directly to the Server using simple HTTP
GET/POST requests.  Using this module together with the Selenium
Server, you can automatically control any supported browser.

To use this module, you need to have already downloaded and started the
Selenium Server.  (The Selenium Server is a Java application.)

The Selenium Server is available here:

  http://www.openqa.org/selenium-rc/

or on CPAN in the Alien-SeleniumRC package:

  http://search.cpan.org/dist/Alien-SeleniumRC

To install:

    perl Makefile.PL
    make
    make test
    make install

Once installed, your test script will look something like this:

    #!/usr/bin/perl
    use strict;
    use warnings;
    use Test::More tests => 4;
    use Test::WWW::Selenium;

    my $sel = Test::WWW::Selenium->new( host => "localhost",    # where selenium-server is running
                                        browser => "*firefox",  # browser to server should launch
                                        browser_url => "http://your.site");

    $sel->open("http://your.site");
    $sel->title_is('Super Cool Mega Site!');
    $sel->text_is('todays_awesomeness', '100');
    $sel->text_like('status', qr/returned 42 results/);


Copyright (c) 2006 Luke Closs <lukec@cpan.org>.
Copyright (c) 2006 Dan Fabulitch <dfabulich@warpmail.net>.
Copyright (c) 2005, 2006 Mattia Barbon <mbarbon@cpan.org>.

Other contributors: Dan Dascalescu, Scott McWhirter.

lib/WWW/Selenium.pm comes from the Selenium-RC project, which releases
it under the Apache License, Version 2.0.

  http://www.apache.org/licenses/LICENSE-2.0

This package is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.



Changes
Revision history for Perl extension Test::WWW::Selenium.
1.36 - Wed May 15 17:38:12 EDT 2013
 - updates dist.ini to hide util with MetaNoIndex


1.35 - Fri Jan  4 14:48:28 EST 2013
 - Updates t/00-compile.t to ignore loading the Devel::REPL plugin, as its an
   optional dependency.

1.34 - Tue Oct 30 13:22:59 EDT 2012
 - Re-release using EU::MM to avoid Module::Install bug
   see http://weblog.bulknews.net/post/33907905561/do-not-ship-modules-with-module-install-1-04
   for details.

1.32 - Tue Feb  7 19:55:00 PST 2012
 - Make Devel::REPL a runtime requirement - thx haarg++

1.31 - Wed Dec  7 22:59:01 EST 2011
  - pod parsing fix for metacpan.

1.30 - Wed Dec  7 22:59:01 EST 2011
  - Added ->debug(), an interactive debugger.

1.25 - Wed Apr 27 23:47:49 EDT 2011
  - Bumped WWW::Selenium version - thanks Leah K.
  - Localize $@ during destruction so we do not clobber it. Reported by Jason S
  - Testsuite fully passing again with latest SeleniumRC (1.0.3)

1.24 - Mon Feb  7 10:28:32 PST 2011
  - clean up pod after nudge from Dave Cross
  - fix synopsis to set the test name correctly - thanks Herbert Leitz

1.23 - 2010-09-05
  - 3 Patches from Tom Hukins++ to clean up inc/, pod and the license.
  - Suggestion from So Matsui++ to restore HTTP POST from the 1.20 release.
  - http_method now defaults to POST but can be set to GET.

1.22 - Fri Apr  2 10:51:58 PDT 2010
  - Update URI::Encode dependency so we can depend on a consistent
    encoding of *firefox to %2Afirefox.
  - No code changes, just test fixes.

1.21 - Wed Dec 30 14:58:07 PST 2009
  - Add error_callback() patch from Derek Wueppelmann

1.20 - Mon Nov  2 22:13:47 PST 2009
  - Refactor timeout code so the UserAgent timeout is always long enough
  - Make set_timeout() also set the UserAgent timeout.

1.19 - Thu Sep 24 19:27:51 PDT 2009
  - Fixed test to not spew deprecation errors
  - removed t/test_ajax_jsf.t because the site it used went away

1.18 - Mon Aug 31 23:20:21 PDT 2009
  - Update list of commands that do not require a locator (barbie++)
  - Update Module::Install to 0.91
  - Ensure that do_command reuses LWP objects and caches 
    connections for performance (konobi++)
  - Minor POD touchups by (dandv++)

1.17 - Tue 30 Dec 2008 00:12:53 PST
 - Allow no_locator methods to be subclassable
   - Thanks to Michael Hendricks of Grant Street Group.

1.16 - Sun 14 Dec 2008 15:28:38 PST
 - use POST instead of GET to selenium server (gyrm)
 - add support for per-session javascript (gyrm)
 - add unit tests and docs for re-using session_ids
 - added LWP::UserAgent dependency to fix cpan testers failures

1.15 - Sun 23 Mar 2008 14:02:25 PDT
 - allow set_timeout() to be called before open()

1.14 - Tue Dec 11 12:45:34 PST 2007
 - fixed typo in Test::WWW::Selenium SYNOPSIS - thanks, Paul Puse
 - added missing dependency on URI::Escape - thanks, Peter (Stig) Edwards

1.13 - Wed Jun  6 19:52:18 PDT 2007
 - fix bug to allow directly calling do_command - Thanks, Tom Hukins
 - Changes from Kevin Jones at Socialtext:
   - refactored out how functions are created
   - add functions to more closely match Selenium IDE:
     - wait_for_text_present
     - wait_for_element_present, and
     - pause

1.12 - Tue Mar 13 20:26:10 PDT 2007
 - move unit test utility to WWW::Selenium::Util
 - refactor unit tests to use Test::Mock::LWP
 - fix bug in get_boolean_array - thanks to Ilya Averyanov

1.11 - Thu Jan  4 19:53:18 PST 2007
 - don't stop a stopped session
 - don't open multiple browsers when start() is called several times
 - Update with Selenium-RC 0.91

1.10 - Tue Nov 14 20:51:13 PST 2006
 - Integrated latest functions available in Selenium-Core 0.90
   - WWW/Selenium.pm is build directly from Core's iedoc.xml
   - Module is built as part of Selenium-RC nightly build process
 - Retained backwards compatibility for deprecated core functions
 - Cleaned up POD
 - Check that caller has opened a page before running commands
 - Open without args opens '/'
 - Added auto_stop option (default on) to auto-close browser windows
 - Remove old get_location function in favour of core builtin
   get_location.  It returns the absolute location.

1.04 - Fri Oct  6 19:02:30 PDT 2006
 - Happy Canadian Thanksgiving!  I thank everyone who uses this.
 - Documentation fixes from Frederic Buclin++
 - Fixed WWW::Selenium example code
 - Patch received from Paul Johnson++ improves:
   - Error message for undefined methods
   - Default test names
   - whitespace and typo fixes
 - Made default_names default to on
 - Removed docs suggesting support for Java-style method names
   (This hasn't worked since < 0.20)

1.03 - Tue Jun 27 22:47:43 PDT 2006
 - Unit tests don't test stderr using Test::Builder::Tester
 - When the Test-Simple folks get T::B::Tester working better I'll 
   re-enable these tests

1.02 - Sat Jun 24 17:25:35 PDT 2006
 - Added unit tests for all auto-generated functions
 - fixed bugs where line numbers and file info wasn't reported
 - fixed bug where test description was lost
 - added get_location method to return relative location
 - removed references to contains and lacks functions
 - failed action_ok() doesn't die anymore, the test just fails

1.01 - Thu Jun 22 13:57:33 PDT 2006
 - No longer call WWW::Selenium::AUTOLOAD if method can't be 
   found (rt bug #20042)
 - Update sample Test::WWW::Selenium code to actually work 
   (Thanks, Shane Landrum)

1.00 - Mon Jun 12 08:53:18 PDT 2006
 - This works well, so make it a stable 1.0 release
 - Better fix for Encode.pm support (Kevin Jones)
 - fixed unit tests when run under 5.8
 - Fixed pod errors in WWW::Selenium

0.24 - Sat May 27 18:08:17 PDT 2006
    - add version to WWW::Selenium
    - fixed Encode.pm support (Kevin Jones)
    - fixed typo (Kevin Jones)

0.22 - Sat May 27 18:08:17 PDT 2006
    - Integrate Dan Fabulitch's latest lib/WWW/Selenium.pm
    - made Encode dependency optional

t/WWW/Selenium.pm
package t::WWW::Selenium;
use strict;
use warnings;
use Test::More;
use Test::Exception;
use Test::Mock::LWP;
use base 'WWW::Selenium';

my $ua_timeout = 180;

sub new {
    my $class = shift;
    my %opts = (
        host => 'localhost',
        port => 4444,
        browser => '*firefox',
        browser_url => 'http://example.com',
        no_deprecation_msg => 1,
        @_,
    );
    my $self = $class->SUPER::new( %opts );

    # Store mock www user agent and startup a session
    $self->_set_mock_response_content('FAKE_SESSION_ID');
    $self->start;

    $self->ua->mock( timeout => sub {
            $ua_timeout = $_[1] if $_[1];
            return $ua_timeout;
        }
    );

    # Test that the session was started as we expect
    my $req_args = $Mock_req->new_args;
    my $url = "http://$opts{host}:$opts{port}/selenium-server/driver/";
    my $content = "cmd=getNewBrowserSession&1=%2Afirefox"
                . "&2=http%3A%2F%2Fexample.com";
    is $req_args->[1], 'POST';
    is $req_args->[2], $url;
    is $req_args->[4], $content;

    return $self;
}

sub _set_mock_response_content {
    my ($self, $content) = @_;
    my $msg = $content;
    if (length($msg) == 0 or $msg !~ /^ERROR/) {
        $msg = "OK,$msg";
    }
    $Mock_resp->mock( content => sub { $msg } );
}

sub _method_exists {
    my ($self, $method, $return_type) = @_;
    my $response = 'Something';
    $response = 'true' if $method =~ m/^(?:is_|get_whether)/i;
    $self->_set_mock_response_content($response);
    lives_ok { $self->$method(1, 2) } "$method lives";
}
1;


Makefile.PL
#!/usr/bin/perl -w
use warnings;
use strict;
use inc::Module::Install 0.95;

# Need to explicitly pre-load this so it doesn't conflict with a mocked one
# created by loading Test::Mock::LWP
eval "require LWP::UserAgent";

name            'Test-WWW-Selenium';
license         'other';
all_from        'lib/Test/WWW/Selenium.pm';
author          'Luke Closs <test-www-selenium@awesnob.com>';
requires       'LWP::UserAgent'  => 0;
requires       'URI::Escape'     => '3.30';
requires       'Devel::REPL'     => 0;
build_requires 'Test::More'      => '0.42';
build_requires 'Test::Exception' => 0;
build_requires 'Test::Mock::LWP' => 0;
build_requires 'Test::Pod'       => 0;

# Auto-install all dependencies from CPAN
auto_install;

# Generate the Makefile
repository 'http://github.com/lukec/cpan-selenium-rc-perl';
WriteAll;


util/fetch_iedoc_xml.pl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::Simple;
use Archive::Zip qw(:ERROR_CODES :CONSTANTS);
use File::Copy qw/copy/;
use File::Path qw/rmtree/;

my $selenium_core_url = 'http://release.openqa.org/selenium-core/1.0.1/selenium-core-1.0.1.zip';
my $core_dir = fetch_and_extract($selenium_core_url);
my $core_iedoc = "$core_dir/core/iedoc.xml";
die "Can't find $core_iedoc" unless -e $core_iedoc;

rmtree 'target';
mkdir 'target';
my $iedoc_dest = 'target/iedoc.xml';
print "Copying $core_iedoc to $iedoc_dest...\n";
copy($core_iedoc => $iedoc_dest) or die "Can't copy $core_iedoc to ${iedoc_dest}: $!";
exit;


sub fetch_and_extract {
    my $url = shift;
    my $tmp_dir = "core-$$";
    mkdir $tmp_dir or die "Can't mkdir $tmp_dir: $!";
    chdir $tmp_dir or die "Can't chdir $tmp_dir: $!";
    (my $zip_file = $url) =~ s#.+/##;
    unless (-e $zip_file) {
            print "Fetching $url...\n";
            getstore($url, $zip_file);
            die "Couldn't fetch $url!" unless -e $zip_file;
    }

    print "Reading $zip_file...\n";
    my $zip = Archive::Zip->new;
    unless ($zip->read($zip_file) == AZ_OK) {
            die "Failed to read $zip_file";
    }

    my $src_dir = $tmp_dir;
    print "Extracting to $src_dir...\n";
    $zip->extractTree;
    chdir '..';
    return $src_dir;
}


lib/WWW/Selenium/Util.pm
package WWW::Selenium::Util;
# ABSTRACT: Utility code to help test using Selenium

use strict;
use warnings;
use IO::Socket;
use base 'Exporter';
our @EXPORT_OK = qw(server_is_running);

=head1 NAME

WWW::Selenium::Util - Utility code to help test using Selenium

=head1 SYNOPSIS

WWW::Selenium::Util contains utility functions to help use Selenium
in your test scripts:

  use WWW::Selenium::Util qw/server_is_running/;
  use Test::More;

  if (server_is_running) {
      plan tests => 1;
  }
  else {
      plan skip_all => "No selenium server found!";
      exit 0;
  }

  # ... your tests ...

=head1 FUNCTIONS

=head2 server_is_running( $host, $port )

Returns true if a Selenium server is running.  The host and port 
parameters are optional, and default to C<localhost:4444>.

Environment vars C<SRC_HOST> and C<SRC_PORT> can also be used to
determine the server to check.

=cut

sub server_is_running {
    my $host = $ENV{SRC_HOST} || shift || 'localhost';
    my $port = $ENV{SRC_PORT} || shift || 4444;

    return ($host, $port) if IO::Socket::INET->new(
        PeerAddr => $host,
        PeerPort => $port,
    );
    return;

}

1;

__END__

=head1 AUTHORS

Written by Luke Closs <selenium@5thplane.com>

=head1 LICENSE

Copyright (c) 2007 Luke Closs <lukec@cpan.org>

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.


lib/Devel/REPL/Plugin/Selenium.pm
package Devel::REPL::Plugin::Selenium;
 
use Devel::REPL::Plugin;
use Data::Dumper;
use namespace::clean -except => [ 'meta' ];
 
has 'selenium' => (isa     => 'Object', is => 'rw');

sub selenium_lex_env {
    my $self = shift;

    my $class = ref($self->selenium);
    my $code = "my \$sel;\n" . Data::Dumper->Dump([$self->selenium], ['sel']);
    $code .= "delete \$sel->{_ua};\n";
    $code .= "use $class; bless \$sel, '$class';";
    return $code;
}

around 'read' => sub {
    my $orig = shift;
    my ($self, @args) = @_;
    my $line = $self->$orig(@args);
    return $line unless $line =~ m/\S/;

    if ($line =~ m/^js:\s*(.+)/) {
        $line = "get_eval(q|$1|)";
    }
    $line =~ s/^(jQuery|\$)\((.+)/get_eval(q|window.jQuery($1)|)/;
    $line = "\$sel->$line";
    warn $line;
    return $line;
};

1;


util/create_www_selenium.pl
#!/usr/bin/perl
use strict;
use warnings;
use lib 'util';
use WWW::Selenium::XMLParseUtils qw/extract_functions html2pod camel2perl
                                    strip_blockquotes create_function/;

my $iedoc = read_iedoc_xml();

# Convert HTML formatting into POD
$iedoc =~ s{^<\?.+?<top>}{#}s; # strip the xml header

$iedoc =~ s#\n##smg; # newlines shouldn't matter to them, but they do matter to us
$iedoc = html2pod($iedoc);

$iedoc = strip_blockquotes($iedoc);

# remove extra newlines
$iedoc =~ s#\n{2,}#\n\n#g;

$iedoc =~ s#^(.+?)</top>##s;
my $selenium_description = $1;

my $function_extras = {
    open => { extra_code => <<'EOT' },
    $_[0] ||= '/'; # default to opening site root
EOT
    set_timeout => { extra_code => <<'EOT' },
    $self->{_timeout} = $_[0];
EOT
    wait_for_pop_up => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
    wait_for_condition => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
    wait_for_page_to_load => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
    wait_for_frame_to_load => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
    wait_for_text_present => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
    wait_for_element_present => { extra_code => <<'EOT' },
    local $self->{_timeout} = $_[1];
EOT
};
my @functions = extract_functions($iedoc, $function_extras);

my @extra_cmds = (
    {
        name => 'waitForTextPresent',
        params => '$text, $timeout',
        desc => 'Waits until $text is present in the html source',
    },
    {
        name => 'waitForElementPresent',
        params => '$locator, $timeout',
        desc => 'Waits until $locator is present',
    },
);

# Print Selenium.pm
write_file("lib/WWW/Selenium.pm", join('', pm_header(),
                                           $selenium_description,
                                           pm_constructor(),
                                           map({ $_->{text} } @functions),
                                           map({create_function(%$_)} @extra_cmds),
                                           pm_footer() ) );

# Print unit test file
my @perl_names = map({camel2perl($_->{name})} @functions, @extra_cmds);
write_file("t/selenium-core.t", join('', t_header(), test_functions(@perl_names) ));
exit;

sub pm_header {
    return <<'EOT';
# Copyright 2006 ThoughtWorks, Inc
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


package WWW::Selenium;
use LWP::UserAgent;
use HTTP::Request;
use HTTP::Headers;
use URI::Escape;
use Carp qw(croak);
use Time::HiRes qw(sleep);

use strict;
use warnings;

our $VERSION = '0.93';

=head1 NAME

WWW::Selenium - Perl Client for the Selenium Remote Control test tool

=head1 SYNOPSIS

    use WWW::Selenium;

    my $sel = WWW::Selenium->new( host => "localhost",
                                  port => 4444,
                                  browser => "*iexplore",
                                  browser_url => "http://www.google.com",
                                );

    $sel->start;
    $sel->open("http://www.google.com");
    $sel->type("q", "hello world");
    $sel->click("btnG");
    $sel->wait_for_page_to_load(5000);
    print $sel->get_title;
    $sel->stop;

=head1 DESCRIPTION

Selenium Remote Control (SRC) is a test tool that allows you to write
automated web application UI tests in any programming language against
any HTTP website using any mainstream JavaScript-enabled browser.  SRC
provides a Selenium Server, which can automatically start/stop/control
any supported browser. It works by using Selenium Core, a pure-HTML+JS
library that performs automated tasks in JavaScript; the Selenium
Server communicates directly with the browser using AJAX
(XmlHttpRequest).

L<http://www.openqa.org/selenium-rc/>

This module sends commands directly to the Server using simple HTTP
GET/POST requests.  Using this module together with the Selenium
Server, you can automatically control any supported browser.

To use this module, you need to have already downloaded and started
the Selenium Server.  (The Selenium Server is a Java application.)

=cut

### This next part is auto-generated based on the big comment in selenium-api.js

EOT
}

sub pm_constructor {
    return <<'EOT';

=cut

eval 'require Encode';
my $encode_present = !$@;
Encode->import('decode_utf8') if $encode_present;

=head2 METHODS

The following methods are available:

=over

=item $sel = WWW::Selenium-E<gt>new( %args )

Constructs a new C<WWW::Selenium> object, specifying a Selenium Server
host/port, a command to launch the browser, and a starting URL for the
browser.

Options:

=over

=item * C<host>

host is the host name on which the Selenium Server resides.

=item * C<port>

port is the port on which the Selenium Server is listening.

=item * C<browser_url>

browser_url is the starting URL including just a domain name.  We'll
start the browser pointing at the Selenium resources on this URL,
e.g. "http://www.google.com" would send the browser to
"http://www.google.com/selenium-server/SeleneseRunner.html"

=item * C<browser> or C<browser_start_command>

This is the command string used to launch the browser, e.g.
"*firefox", "*iexplore" or "/usr/bin/firefox"

This option may be any one of the following:

=over

=item * C<*firefox [absolute path]>

Automatically launch a new Firefox process using a custom Firefox
profile.
This profile will be automatically configured to use the Selenium
Server as a proxy and to have all annoying prompts
("save your password?" "forms are insecure" "make Firefox your default
browser?" disabled.  You may optionally specify
an absolute path to your firefox executable, or just say "*firefox".
If no absolute path is specified, we'll look for
firefox.exe in a default location (normally c:\program files\mozilla
firefox\firefox.exe), which you can override by
setting the Java system property C<firefoxDefaultPath> to the correct
path to Firefox.

=item * C<*iexplore [absolute path]>

Automatically launch a new Internet Explorer process using custom
Windows registry settings.
This process will be automatically configured to use the Selenium
Server as a proxy and to have all annoying prompts
("save your password?" "forms are insecure" "make Firefox your default
browser?" disabled.  You may optionally specify
an absolute path to your iexplore executable, or just say "*iexplore".
If no absolute path is specified, we'll look for
iexplore.exe in a default location (normally c:\program files\internet
explorer\iexplore.exe), which you can override by
setting the Java system property C<iexploreDefaultPath> to the correct
path to Internet Explorer.

=item * C</path/to/my/browser [other arguments]>

You may also simply specify the absolute path to your browser
executable, or use a relative path to your executable (which we'll try
to find on your path).  B<Warning:> If you
specify your own custom browser, it's up to you to configure it
correctly.  At a minimum, you'll need to configure your
browser to use the Selenium Server as a proxy, and disable all
browser-specific prompting.

=back

=item * C<auto_stop>

Defaults to true, and will attempt to close the browser if the object
goes out of scope and stop hasn't been called.

=item * C<keep_alive>

Number of connections LWP should cache. This is just a minor speed
improvement. Defaults to 5.

=item * C<http_method>>

Choose which HTTP method should be used for requests to the Selenium server.
Only GET and POST are supported.

=back

=cut

sub new {
    my ($class, %args) = @_;
    my $self = { # default args:
                 host => 'localhost',
                 port => 4444,
                 auto_stop => 1,
                 browser_start_command => delete $args{browser} || '*firefox',
                 _ua => undef,
                 keep_alive => 5,
                 http_method => 'POST',
                 %args,
               };
    croak 'browser_url is mandatory!' unless $self->{browser_url};
    croak "Unknown http_method: ($self->{http_method})"
        unless $self->{http_method} =~ m/^GET|POST$/;
    bless $self, $class or die "Can't bless $class: $!";
    return $self;
}

sub start {
    my $self = shift;
    return if $self->{session_id};
    $self->{session_id} = $self->get_string("getNewBrowserSession",
                                            $self->{browser_start_command},
                                            $self->{browser_url});
}

sub stop {
    my $self = shift;
    return unless defined $self->{session_id};
    $self->do_command("testComplete");
    $self->{session_id} = undef;
}

sub do_command {
    my ($self, $command, @args) = @_;
    my $get = $self->{http_method} eq 'GET';

    $self->{_page_opened} = 1 if $command eq 'open';

    # Check that user has called open()
    my %valid_pre_open_commands = (
        testComplete => 1,
        getNewBrowserSession => 1,
        setTimeout => 1,
    );
    if (!$self->{_page_opened} and !$valid_pre_open_commands{$command}) {
        die "You must open a page before calling $command. eg: \$sel->open('/');\n";
    }

    my $fullurl = "http://$self->{host}:$self->{port}/selenium-server/driver/";
    $fullurl .= '?' if $get;
    my $content = '';

    my $i = 1;
    @args = grep defined, @args;
    my $params = $get ? \$fullurl : \$content;
    $$params .= "cmd=" . uri_escape($command);
    while (@args) {
        $$params .= '&' . $i++ . '=' . URI::Escape::uri_escape_utf8(shift @args);
    }
    if (defined $self->{session_id}) {
        $$params .= "&sessionId=$self->{session_id}";
    }
    # We use the full version of LWP to make sure we issue an
    # HTTP 1.1 request (SRC-25)

    my $method = $get ? 'GET' : 'POST';
    print "---> Requesting $method $fullurl ($content)\n" if $self->{verbose};
    my $header = HTTP::Headers->new(
        $get ? () : (
            Content_Type => 'application/x-www-form-urlencoded; charset=utf-8'
        )
    );
    my $response = $self->ua->request(
        HTTP::Request->new($method => $fullurl, $header, $content) );
    my $result;
    if ($response->is_success) {
        $result = $response->content;
        print "Got result: $result\n" if $self->{verbose};
    }
    else {
        die "Error requesting $fullurl:\n" . $response->status_line . "\n";
    }
    $result = decode_utf8($result) if $encode_present;
    die "Error requesting $fullurl:\n$result\n" unless $result =~ /^OK/;
    return $result;
}

sub ua {
    my $self = shift;

    $self->{_ua} ||= LWP::UserAgent->new(keep_alive => $self->{keep_alive});

    if (my $msec = $self->{_timeout}) {
        # Keep the 3 minute timeout (LWP::UserAgent default) on top of the
        # selenium timeout
        $self->{_ua}->timeout( int($msec/1000 + 180) );
    }
    return $self->{_ua};
}

sub get_string {
    my $self = shift;
    my $result = $self->do_command(@_);
    return substr($result, 3);
}

sub get_string_array {
    my $self = shift;
    my $result = $self->get_string(@_);
    my $token = "";
    my @tokens = ();
    my @chars = split(//, $result);
    for (my $i = 0; $i < @chars; $i++) {
        my $char = $chars[$i];
        if ($char eq '\\') {
            $i++;
            $char = $chars[$i];
            $token .= $char;
        } elsif ($char eq ',') {
            push (@tokens, $token);
            $token = "";
        } else {
            $token .= $char;
        }
    }
    push (@tokens, $token);
    return @tokens;
}

sub get_number {
    my $self = shift;
    my $result = $self->get_string(@_);
    # Is there something else I need to do here?
    return $result;
}

sub get_number_array {
    my $self = shift;
    my @result = $self->get_string_array(@_);
    # Is there something else I need to do here?
    return @result;
}

sub get_boolean {
    my $self = shift;
    my $result = $self->get_string(@_);
    if ($result eq "true") {
        return 1;
    }
    if ($result eq "false") {
        return 0;
    }
    die "result is neither 'true' nor 'false': $result";
}

sub get_boolean_array {
    my $self = shift;
    my @result = $self->get_string_array(@_);
    my @boolarr = ();
    for (my $i = 0; $i < @result; $i++) {
        if ($result[$i] eq "true") {
            push (@boolarr, 1);
            next;
        }
        if ($result[$i] eq "false") {
            push (@boolarr, 0);
            next;
        }
        die "result is neither 'true' nor 'false': ". $result[$i];
    }
    return @boolarr;
}

=item $sel-E<gt>pause($timeout)

Waits $timeout milliseconds (default: 1 second)

=cut

sub pause {
    my ($self,$timeout) = @_;
    $timeout = 1000  unless defined $timeout;
    $timeout /= 1000;
    sleep $timeout;
}

### From here on, everything's auto-generated from XML

EOT
}

sub pm_footer {
    return <<'EOT';

=item * $sel-E<gt>is_location($expected_location)

Verify the location of the current page ends with the expected location.
If an URL querystring is provided, this is checked as well.

$expected_location is the location to match.

Note: This function is deprecated, use get_location() instead.

=cut

sub is_location {
    my $self = shift;
    warn "is_location() is deprecated, use get_location()\n"
        unless $self->{no_deprecation_msg};
    my $expected_location = shift;
    my $loc = $self->get_string("getLocation");
    return $loc =~ /\Q$expected_location\E$/;
}

=item * $sel-E<gt>get_checked($locator)

Gets whether a toggle-button (checkbox/radio) is checked.  Fails if the specified element doesn't exist or isn't a toggle-button.

$locator is an element locator pointing to a checkbox or radio button.

Note: This function is deprecated, use is_checked() instead.

=cut

sub get_checked {
    my $self = shift;
    warn "get_checked() is deprecated, use is_checked()\n"
        unless $self->{no_deprecation_msg};
    return $self->get_string("isChecked", @_) ? 'true' : 'false';
}

=item * $sel-E<gt>is_selected($locator, $option_locator)

Verifies that the selected option of a drop-down satisfies the optionSpecifier.

See the select command for more information about option locators.

$locator is an element locator.
$option_locator is an option locator, typically just an option label (e.g. "John Smith").

Note: This function is deprecated, use the get_selected_*() methods instead.

=cut

sub is_selected {
    my ($self, $locator, $option_locator) = @_;
    warn "is_selected() is deprecated, use get_selected_*() methods\n"
        unless $self->{no_deprecation_msg};
    $option_locator =~ m/^(?:(.+)=)?(.+)/;
    my $selector = $1 || 'label';
    $selector = 'indexe' if $selector eq 'index';
    my $pattern = $2;
    my $func = "get_selected_${selector}s";
    my @selected = $self->$func($locator);
    return grep { $pattern eq $_ } @selected;
}

=item * $sel-E<gt>get_selected_options($locator)

Gets all option labels for selected options in the specified select or multi-select element.

$locator is an element locator.

Note: This function is deprecated, use get_selected_labels() instead.

=cut

sub get_selected_options {
    my $self = shift;
    warn "get_selected_options() is deprecated, use get_selected_labels()\n"
        unless $self->{no_deprecation_msg};
    return $self->get_string_array("getSelectedLabels", @_);
}

=item * $sel-E<gt>get_absolute_location()

Gets the absolute URL of the current page.

Note: This function is deprecated, use get_location() instead.

=cut

sub get_absolute_location {
    my $self = shift;
    warn "get_absolute_location() is deprecated, use get_location()\n"
        unless $self->{no_deprecation_msg};
    return $self->get_string("getLocation", @_);
}

=pod

=back

=cut

sub DESTROY {
    my $self = shift;
    $self->stop if $self->{auto_stop};
}

1;

__END__

=head1 SEE ALSO

For more information about Selenium Remote Control, visit the website
at L<http://www.openqa.org/selenium-rc/>.

=head1 BUGS

The Selenium Remote Control JIRA issue tracking system is available
online at L<http://jira.openqa.org/browse/SRC>.

=head1 AUTHOR

Perl driver maintained by Luke Closs <selenium-rc@awesnob.com>

Selenium Remote Control maintained by Dan Fabulich <dfabulich@warpmail.net>

=head1 LICENSE

Copyright (c) 2006 ThoughtWorks, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

EOT
}

sub t_header {
    return <<'EOT';
#!/usr/bin/perl
use strict;
use warnings;
use Test::More qw/no_plan/;

BEGIN {
    use lib 't/lib';
    use_ok 'LWP::UserAgent';    # mocked
    use_ok 'HTTP::Response';    # mocked
    use lib 'lib';
    use t::WWW::Selenium;
}

my $sel = t::WWW::Selenium->new;
isa_ok $sel, 't::WWW::Selenium';
$sel->open;

EOT
}

sub test_functions {
    my @funcs = @_;
    my $text = '';
    for my $f (@funcs) {
        $text .= qq{\$sel->_method_exists("$f");\n};
    }
    return $text;
}

sub read_iedoc_xml {
    my $iedoc_file = 'target/iedoc.xml';
    die "Can't find iedoc.xml" unless -e $iedoc_file;
    open (my $fh, $iedoc_file) or die "Can't open $iedoc_file: $!";
    {
        local $/ = undef;
        $iedoc = <$fh>;
    }
    close $fh;
    return $iedoc;
}

sub write_file {
    my ($name, $content) = @_;
    open(my $fh, ">$name") or die "Can't open $name: $!";
    print $fh $content;
    close $fh or die "Can't write $name: $!";
}
